File: library/src/main/java/nz/co/cic/wifi/Station.java
package nz.co.cic.wifi;

import android.content.Context;
import android.net.wifi.p2p.WifiP2pGroup;
import android.net.wifi.p2p.WifiP2pManager;
import android.os.Looper;

import org.jdeferred.Deferred;
import org.jdeferred.DoneCallback;
import org.jdeferred.FailCallback;
import org.jdeferred.Promise;
import org.jdeferred.impl.DeferredObject;

/**
 * Created by dipshit on 3/03/17.
 */

public class Station {

    private WifiP2pManager p2pManager;
    private WifiP2pManager.Channel channel;
    private Context context;

    public Station(Context c){
        this.context = c;
        this.p2pManager = (WifiP2pManager) c.getSystemService(Context.WIFI_P2P_SERVICE);
        this.channel = p2pManager.initialize(this.context, Looper.getMainLooper(), new WifiP2pManager.ChannelListener() {
            @Override
            public void onChannelDisconnected() {
                //TODO handle disconnect
            }
        });
    }

    public void start(final Deferred deferred){
        Deferred deferred1 = new DeferredObject();
        Promise p = deferred1.promise();
        p.done(new DoneCallback() {
            @Override
            public void onDone(Object result) {
                getInfo(deferred);
            }
        }).fail(new FailCallback() {
            @Override
            public void onFail(Object result) {
                deferred.reject(result);
            }
        });

        startRadio(deferred1);
    }

    public void stop(Deferred deferred){
        stopRadio(deferred);
    }

    public void getInfo(final Deferred deferred){
        this.p2pManager.requestGroupInfo(channel, new WifiP2pManager.GroupInfoListener() {
            @Override
            public void onGroupInfoAvailable(WifiP2pGroup group) {
                if(group != null){
                    deferred.resolve(group);
                }else{
                    deferred.reject("No group available");
                }
            }
        });
    }

    private void startRadio(final Deferred deferred){
        this.p2pManager.createGroup(channel, new WifiP2pManager.ActionListener() {
            @Override
            public void onSuccess() {
                deferred.resolve(null);
            }
            @Override
            public void onFailure(int reason) {
                deferred.reject(reason);
            }
        });
    }

    private void stopRadio(final Deferred deferred){
        this.p2pManager.removeGroup(channel, new WifiP2pManager.ActionListener() {
            @Override
            public void onSuccess() {
                deferred.resolve(null);
            }

            @Override
            public void onFailure(int reason) {
                deferred.reject(reason);
            }
        });
    }


}

File: library/src/main/java/nz/co/cic/wifi/Station.kt
package nz.co.cic.wifi

import android.content.Context
import android.net.wifi.p2p.WifiP2pGroup
import android.net.wifi.p2p.WifiP2pManager
import android.os.Looper
import io.reactivex.BackpressureStrategy
import io.reactivex.Flowable
import io.reactivex.FlowableEmitter


/**
 * Created by dipshit on 3/03/17.
 */

class Station(private val context: Context) {

    private val p2pManager: WifiP2pManager
    private val channel: WifiP2pManager.Channel

    init {
        this.p2pManager = context.getSystemService(Context.WIFI_P2P_SERVICE) as WifiP2pManager
        this.channel = p2pManager.initialize(this.context, Looper.getMainLooper()) {
            //TODO handle disconnect
        }
    }

    fun start(): Flowable<WifiP2pGroup>{
        return Flowable.create({
            subscriber ->
            startRadio().subscribe({}, {
                err ->
                subscriber.onError(err)
            }, {
                getInfo(subscriber)
            })

        }, BackpressureStrategy.BUFFER)
    }

    fun stop(): Flowable<Boolean>{
        return stopRadio()
    }

    fun getInfo(emitter: FlowableEmitter<WifiP2pGroup>) {
        this.recursiveGetInfo(emitter)
    }

    private fun recursiveGetInfo(emitter: FlowableEmitter<WifiP2pGroup>){
        this._getInfo(0, 5, emitter)
    }

    private fun _getInfo(attempt: Int, limit: Int, emitter: FlowableEmitter<WifiP2pGroup>){
        this.p2pManager.requestGroupInfo(channel) { group ->
            if (group != null) {
                emitter.onNext(group)
            }else if(group == null && attempt < limit){
                Thread.sleep(200)
                var tmp_attempt = attempt.plus(1)
                this._getInfo(tmp_attempt, limit, emitter)
            }else {
                emitter.onError(Throwable("No group available"))
            }
        }
    }

    private fun startRadio() : Flowable<Boolean> {
        return Flowable.create({
            subscriber ->
            this.p2pManager.createGroup(channel, object : WifiP2pManager.ActionListener {
                override fun onSuccess() {
                   subscriber.onComplete()
                }

                override fun onFailure(reason: Int) {
                    if(reason == WifiP2pManager.P2P_UNSUPPORTED){
                        subscriber.onError(Throwable("P2P UNSUPPORTED"))
                    }else if(reason == WifiP2pManager.ERROR){
                        subscriber.onError(Throwable("P2P INTERNAL ERROR"))
                    }else{
                        subscriber.onComplete()
                    }
                }
            })

        }, BackpressureStrategy.BUFFER)

    }

    private fun stopRadio(): Flowable<Boolean>{
        return Flowable.create({
            subscriber ->

            this.p2pManager.removeGroup(channel, object : WifiP2pManager.ActionListener {
                override fun onSuccess() {
                    subscriber.onComplete()
                }

                override fun onFailure(reason: Int) {
                    subscriber.onError(Throwable("Error: " + reason))
                }
            })
        }, BackpressureStrategy.BUFFER)

    }


}
